Fixes bugs in stack unwinding metadata (PR#7118,7120)

btj opened the pull request:
This fixes a handful of bugs in ocamlopt's stack unwinding metadata generation. These have been causing crashes in my GUI application on OS X. (Cocoa performs stack walks as part of its normal operation. Incorrect stack unwinding metadata can cause it to try to read unmapped addresses. See PR#7118.) Besides, this also enables proper full stack backtraces in platform debuggers like gdb and lldb.

I added a regression test.

btj committed:
Add missing .cfi_adjust_cfa_offset before .cfi_endproc

Fixes bug 7120.

btj committed:
Updated Changelog

btj committed:
Fixed some bugs in stack unwinding metadata of runtime routines

It's easy to forget that the assembler ignores the program's
control flow when computing the resulting CFA offset of the
various CFI_ADJUST directives. Note that unbalanced
CFI_ADJUST directives in one assembly routine cause all offsets
in all subsequent routines to be shifted by the corresponding
amount.

btj committed:
Enabled full stack trace in platform debuggers

Platform debuggers such as `gdb` and `lldb` can now show a full stack trace for
call stacks that include ocamlopt-generated code. Specifically, this commit adds
base pointer information to the stack unwinding metadata of runtime function
`caml_start_program`.

gasche commented:
cc @xavierleroy @mshinwell 

(As noted in the Changes entry) this is the continuation of [PR#7118](http://caml.inria.fr/mantis/view.php?id=7118) (problem report) and [PR#7120](http://caml.inria.fr/mantis/view.php?id=7120) (problem analysis and fix suggestion).

btj committed:
Add missing unwind info for 'caml_start_program'

This enables correct stack unwinding, and helps prevent random crashes on OS X
due to Cocoa or other Objective-C (or C++) external functions walking
our stack.

btj committed:
Add test case for unwind info

This test case runs only on OS X. It serves as a regression test
for PR#7120.

btj committed:
Add README file motivating unwind test case

btj committed:
Eliminated a bash dependency

btj committed:
Fixed bug in 'clean' script

damiendoligez commented:
I'd like to see this reviewed by a back-end expert.

mshinwell commented:
Acknowledged, I will look at this.
Has this been tested on x86-64 Linux?  If not please do so.

btj commented:
So far, I only tested on OS X 10.10.1. (Travis tested on Linux, though.) I will test on Linux for any kind of regressions that I can think of checking for. But for this it would be useful to know what people are trying to achieve by emitting the CFI information.

btj commented:
Answering my own question: CFI information was originally added on [Feb 21, 2012](https://github.com/ocaml/ocaml/commit/2eecf2d4c0933f64bcfa3e620f031497166db338) in response to [PR#5487](http://caml.inria.fr/mantis/view.php?id=5487), to get decent backtraces in gdb on Linux/x86 and Linux/amd64. I will test that this is still the case.

mshinwell commented:
Right, the CFI information is so the stack can be unwound properly using a DWARF client such as gdb or lldb.  OCaml's internal backtrace infrastructure works from frame descriptors, and is independent.

btj commented:
I built my fork (btj/ocaml) and ran the OCaml test suite on Ubuntu 15.10/amd64. All tests passed. I tested gdb backtrace behavior on a modified version of my regression test program (see [unwind.zip](https://github.com/ocaml/ocaml/files/108577/unwind.zip)); the [backtrace.txt](https://github.com/ocaml/ocaml/files/108581/backtrace.txt) looks fine.

btj committed:
Fixed unwind test case output syntax

mshinwell commented:
@btj Did you test this with the OCaml compiler configured both with and without frame pointers (there is an option to the configure script)?  If not please do so (on both OS X and Linux if possible).

btj commented:
Without frame pointers: see above.
With frame pointers:
- On ubuntu-14.04.3-desktop-amd64: builds fine, OCaml test suite passes, using my [unwind.zip](https://github.com/ocaml/ocaml/files/108577/unwind.zip) example I get a good-looking backtrace 
  [backtrace-with-fp.txt](https://github.com/ocaml/ocaml/files/125489/backtrace-with-fp.txt).
- On OS X 10.10.1: ./configure -with-frame-pointers reports an error, saying that frame pointers are not supported on this platform.

damiendoligez commented:
Thank you for the patch and the tests.

mshinwell commented:
I will finish looking at this shortly

mshinwell commented:
@damiendoligez This patch is OK.  However could you please have a look and see whether the test Makefile is satisfactory?  For the moment this only does anything on OS X, but let's leave it like that for the moment, otherwise we need to add libunwind detection (which I have code for somewhere I can add for 4.04).  However I think it should be using $(OCAMLOPT) rather than just "ocamlopt".  I also wondered whether it should be changed so it doesn't clean itself every time.

@btj Thanks for the work on this.

damiendoligez commented on testsuite/tests/unwind/Makefile:
```diff
@@ -0,0 +1,19 @@
+default:
+	@if $(BYTECODE_ONLY); then : ; elif [ $(SYSTEM) = macosx ]; then $(MAKE) native_macosx_tests; fi
```
This should be something like:

```
    @printf " ... testing 'unwind_test':"
    @if [ $(SYSTEM) = macosx ] && ! $(BYTECODE_ONLY); then \
      $(MAKE) native_macosx_test; \
    else \
      echo " => skipped"; \
    fi
```

... (see below)

damiendoligez commented on testsuite/tests/unwind/Makefile:
```diff
@@ -0,0 +1,19 @@
+	@if $(BYTECODE_ONLY); then : ; elif [ $(SYSTEM) = macosx ]; then $(MAKE) native_macosx_tests; fi
+
+native_macosx_tests:
+	@printf " ... testing 'unwind_test':"; \
```
... and the `printf` should be removed from here.

damiendoligez commented on testsuite/tests/unwind/Makefile:
```diff
@@ -0,0 +1,19 @@
+	./unwind_test >/dev/null 2>&1 && echo " => passed" || echo " => failed"
+
+unwind_test:
+	@ocamlopt -c -opaque mylib.mli
```
@mshinwell is right, these should be `$(OCAMLOPT)` because you can't assume that OCaml is installed.

mshinwell replied:
(And, more importantly, if it is installed it is almost certainly not the compiler you want to test with!)

gasche replied:
(well, [opam-compiler-conf](https://github.com/gasche/opam-compiler-conf) makes it very easy to install a switch for the compiler branch you are currently hacking on.)

mshinwell replied:
The testsuite is always supposed to use the ocamlopt in the source tree, no?

damiendoligez commented:
@btj do you have the time to fix the tests quickly? If not, I'll do it myself and merge manually.

btj committed:
Some fixes in the Makefile for the test case

btj commented:
Done. $(OCAMLOPT): oops; yes, funny that I didn't notice that myself. I must have always run the test suite after installing the build into my PATH.

damiendoligez commented:
On MacOS 10.8, the test fails with this error message:

```
ld: unknown option: -keep_dwarf_unwind
collect2: ld returned 1 exit status
```

The manual page for `ld` mentions `unwind` in only one place: the documentation of  `-warn-compact-unwind`.

I see two possible solutions:
1. Make the test conditional on the MacOS version (somehow).
2. Drop support for MacOS <= 10.8.

@btj which one would you recommend?

btj commented:
I would recommend making the test conditional on the OS X version. I'll try to do that within the next few hours.

btj commented:
Opened pull request #476.
